### PR TITLE
Issue #9973 - use URIUtil.addPaths() instead of URI.resolve()

### LINKED ISSUE #9973: Creating a Resource for an entry in a nested jar file in Jetty 12
**Jetty version**

Jetty 12.0.0.Beta2

**Java version**

17

**Question**

I'm working on upgrading Spring Boot to Jetty 12. Spring Boot supports executable jars and wars that can be run using `java -jar`. This results in URLs pointing to jars nested within a jar or war. For example, this URL taken from our test suite:

```
jar:file:/Users/awilkinson/dev/spring-projects/spring-boot/main/spring-boot-tests/spring-boot-integration-tests/spring-boot-server-tests/build/spring-boot-server-tests-app/build/libs/spring-boot-server-tests-app-jetty.war!/WEB-INF/lib/spring-boot-server-tests-app-resources.jar!/
```

Much of the path doesn't matter. A shorter example would be:

```
jar:file:/example.war!/WEB-INF/lib/spring-boot-server-tests-app-resources.jar!/
```

With Jetty 11 and earlier, I could create a `Resource` that points to the `META-INF/resources` entry within `spring-boot-server-tests-app-resources.jar!` by calling `Resource.newResource(url + "META-INF/resources")`. I am struggling to find the Jetty 12 equivalent of this code.

I have a `WebAppContext` to hand so I have tried calling `ResourceFactory resourceFactory = ResourceFactory.of(handler)` and calling `ResourceFactory.newResource`. This works for situations where `spring-boot-server-tests-app-resources.jar` is available directly on the file system but not for the case where it's nested within a war.

How can I get this working for nested jars using Jetty 12 please?

### PR DESCRIPTION
**Pull request by joakime**

### CODE REVIEW

**Review by gregw — CHANGES_REQUESTED** [review]
Whilst this individual fix looks good, we need to address #10057 and make the URLResource a first class citizen again.   Please compare against the version in jetty-11 and make sure that it is as compatible as possible.

Also it would be good to see if we could get a test that was closer to the spring-boot usage and maybe even use a custom URLStreamHandler?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java [hunk 1]
```diff
@@ -62,18 +90,39 @@ public void setUseCaches(boolean useCaches)
...
+                }
+            }
         }
+        return urlResource;
     }
 
-    private static class URLResource extends Resource
+    private void cleanupReferences()
     {
+        // TODO
```
Comment (gregw) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java]: We should probably do the cleanup before 12.0.0, but perhaps not for beta3.
I was thinking it should be like the JVM's LocaleObjectCache, in that the factory should keep a list of soft references to all resources given out and that we always clean the stale objects on any new resource (which will call close) and then if the factory is closed, we should close all remaining resources, even if still referenced.

**Review by gregw — DISMISSED** [review]
There are still differences with jetty-11, for example, isDurectory checks for existence as well as the trailing '/'.  `getName` is also different?

Please open an issue for the cleanup 

Other than that looks good.

**Comment by gregw** [comment]
We should probably do the cleanup before 12.0.0, but perhaps not for beta3.
I was thinking it should be like the JVM's LocaleObjectCache, in that the factory should keep a list of soft references to all resources given out and that we always clean the stale objects on any new resource (which will call close) and then if the factory is closed, we should close all remaining resources, even if still referenced.

**Comment by joakime** [comment]
This branch makes the spring-boot build green.

The spring-boot branch I'm working with can be found at
https://github.com/joakime/spring-boot-jetty-12/tree/jetty-12-joakim
(complete with code changes to spring-boot)

**Review by lorban — APPROVED** [review]

**Comment by lorban** [comment]
Note that this also solves https://github.com/eclipse/jetty.project/issues/10057

**Pull request merged**